src/Tasks/Microsoft.NET.Build.Tasks/CompileReadyToRunImages.cs

```diff
@@ -0,0 +1,437 @@
+
+        private Architecture _targetArchitecture;
+
+        private bool _crossgenFailures = false;
+
+        public override bool Execute()
+        {
+            // Get the list of runtime identifiers that we support and can target
+            ITaskItem frameworkRef = KnownFrameworkReferences.Where(item => String.Compare(item.ItemSpec, "Microsoft.NETCore.App", true) == 0).SingleOrDefault();
+            string supportedRuntimeIdentifiers = frameworkRef == null ? null : frameworkRef.GetMetadata("RuntimePackRuntimeIdentifiers");
+
+            // Get information on the runtime package used for the current target
+            ITaskItem frameworkPack = RuntimePacks.Where(pack => pack.ItemSpec.EndsWith(".Microsoft.NETCore.App", StringComparison.InvariantCultureIgnoreCase)).SingleOrDefault();
+            _runtimeIdentifier = frameworkPack == null ? null : frameworkPack.GetMetadata(MetadataKeys.RuntimeIdentifier);
+            _packagePath = frameworkPack == null ? null : frameworkPack.GetMetadata(MetadataKeys.PackageDirectory);
+
+            var runtimeGraph = new RuntimeGraphCache(BuildEngine4, Log).GetRuntimeGraph(RuntimeGraphPath);
+            var supportedRIDsList = supportedRuntimeIdentifiers == null ? Array.Empty<string>() : supportedRuntimeIdentifiers.Split(';');
+
+            // Get the best RID for the host machine, which will be used to validate that we can run crossgen for the target platform and architecture
+            string hostRuntimeIdentifier = NuGetUtils.GetBestMatchingRid(
+                runtimeGraph, 
+                DotNet.PlatformAbstractions.RuntimeEnvironment.GetRuntimeIdentifier(), 
+                supportedRIDsList, 
+                out bool wasInGraph);
+
+            if (hostRuntimeIdentifier == null || _runtimeIdentifier == null || _packagePath == null)
+            {
+                Log.LogError(Strings.ReadyToRunNoValidRuntimePackageError);
+                return false;
+            }
+
+            if (!ExtractTargetPlatformAndArchitecture(_runtimeIdentifier, out string targetPlatform, out _targetArchitecture) ||
+                !ExtractTargetPlatformAndArchitecture(hostRuntimeIdentifier, out string hostPlatform, out Architecture hostArchitecture) ||
+                targetPlatform != hostPlatform)
+            {
+                Log.LogError(Strings.ReadyToRunTargetNotSuppotedError);
+                return false;
+            }
+
+            if (!GetCrossgenComponentsPaths() || !File.Exists(_crossgenPath) || !File.Exists(_clrjitPath))
+            {
+                Log.LogError(Strings.ReadyToRunTargetNotSuppotedError);
+                return false;
+            }
+
+            LogStandardErrorAsError = true;
+
+            // Run Crossgen on input assemblies
+            ProcessInputFileList(FilesToPublishPreserveNewest, _r2rPublishPreserveNewest);
+            ProcessInputFileList(FilesToPublishAlways, _r2rPublishAlways);
+
+            return !_crossgenFailures;
```

Comment:
This is one of the most common task authoring bugs when there's a mismatch between Log.HasLoggedErrors and the return value. Usually, we use return !Log.HasLoggedErrors (enforced by our non-tool base class). For tools there's the extra case of non-zero exit code without a captured output that was converted into an error. The base ToolTask handles all of the cases for you, but we're not using it. 

I think this needs to be something like this to get the invariant right on your own:
``` C#
if (_crossgenFailures && !Log.HasLoggedErrors)
{
    Log.LogError(SomethingGeneric);
}

return !Log.HasLoggedErrors;
}

Reply:
Uh, no, our custom Log that was copied over from TaskBase won't know to return true for HasLoggedErrors that came from ExecuteTool. I'll fix this up after this is merged. You can leave this for now.

Reply:
I added a generic error, but getting into the if statement will never happen, given how crossgen has to report a failure message to stderr on errors.